### PR TITLE
Projection connector

### PR DESCRIPTION
**Pull request by nune-tadevosyan**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Adds `MultiLayerProjectionConnector` support for SALM models. 

# Changelog

- Adds support for the `MultiLayerProjectionConnector`.
- Adds support for initializing SALM model from pretrained checkpoints.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/speechlm2/models/salm.py [hunk 1]
```diff
@@ -69,6 +69,22 @@ def __init__(self, cfg) -> None:
         # Load the pretrained streaming ASR model and copy its parameters into the audio perception module.
         setup_speech_encoder(self, pretrained_weights=self.cfg.pretrained_weights)
 
+
+        # Load pretrained weights if provided
```
Comment (pzelasko) [inline nemo/collections/speechlm2/models/salm.py]: Why is this needed in the PR? Looks to me like a separate feature that we may want to implement in a generic way so that every speechlm2 model can re-use.

File: nemo/collections/speechlm2/modules/perception.py [hunk 2]
```diff
@@ -49,6 +49,20 @@ def __init__(self, cfg: DictConfig):
         else:
             self.spec_augmentation = None
         self.modality_adapter = self.from_config_dict(cfg.modality_adapter)
+        if isinstance(self.modality_adapter, (QformerConnector, MultiLayerProjectionConnector)):
+            # Store as a plain Python attribute (not an nn.Module submodule) to avoid duplicate
+            # state_dict entries: encoder_multilayer.encoder IS self.encoder, so safetensors
+            # would de-duplicate them and produce spurious "missing key" warnings on reload.
```
Comment (pzelasko) [inline nemo/collections/speechlm2/modules/perception.py]: It might be an overkill. Can we instead do:
```python
# note: not self.encoder
encoder = self.from_config_dict(cfg.encoder)

if isinstance(self.modality_adapter, (QformerConnector, MultiLayerProjectionConnector)):
    self.encoder_multilayer = ConformerMultiLayerFeatureExtractor(encoder, ...)
else:
    self.encoder = encoder
```

**Review by pzelasko — CHANGES_REQUESTED** [review]

File: nemo/collections/speechlm2/models/salm.py [hunk 3]
```diff
@@ -64,26 +64,10 @@ def __init__(self, cfg) -> None:
...
-            from safetensors import safe_open
-
-            tensors = {}
-            with safe_open(init_from_path / "model.safetensors", framework="pt") as f:
-                for k in f.keys():
-                    tensors[k] = f.get_tensor(k)
-            missing_keys, unexpected_keys = self.load_state_dict(tensors, strict=False)
-            logging.warning(f"Missing keys: {missing_keys}")
-            logging.warning(f"Unexpected keys: {unexpected_keys}")
+        maybe_install_lora(self)
```
Comment (pzelasko) [inline nemo/collections/speechlm2/models/salm.py]: Can we move `maybe_install_lora` back to line 67?

Comment (nune-tadevosyan) [reply nemo/collections/speechlm2/models/salm.py]: Done

File: nemo/collections/speechlm2/modules/perception.py (hunk 2 above)
Comment (nune-tadevosyan) [reply nemo/collections/speechlm2/modules/perception.py]: Done

**Review by pzelasko — APPROVED** [review]

**Comment by pzelasko** [comment]
Force merging - the failing tests are known to be broken on main; they are addressed separately.
